rust/worker/src/segment/distributed_hnsw_segment.rs

```diff
@@ -191,9 +191,37 @@ impl SegmentWriter for DistributedHNSWSegmentWriter {
                         .read()
                         .add(segment_offset_id as usize, &embedding);
                 }
-                Operation::Upsert => {}
-                Operation::Update => {}
-                Operation::Delete => {}
+                Operation::Upsert => {
+                    // hnsw index behavior is to treat add() as upsert
+                    let segment_offset_id = record.0.segment_offset_id;
+                    // Assumption: Upserts must have embedding set
+                    let embedding = record.0.log_record.record.embedding.as_ref().unwrap();
```

Comment:
Can we consider using expect instead of unwrap?

Reply:
Yep good point